File: shared/hermes/hermes/messaging.py
"""Hermes messaging layer initialisation and singleton management."""

from __future__ import annotations

import logging

from pydantic_settings import BaseSettings, SettingsConfigDict

from hermes.interfaces.messaging_service import IMessagingService
from hermes.options import MessagingOptions
from hermes.services.smtp_messaging_service import SmtpMessagingService

logger = logging.getLogger(__name__)

_service: IMessagingService | None = None


class _SmtpEnvConfig(BaseSettings):
    """Internal settings model for loading SMTP configuration from environment variables.

    All field names correspond to environment variables after uppercasing:
    ``smtp_host`` → ``SMTP_HOST``, etc.
    """

    model_config = SettingsConfigDict(
        env_file=None,
        extra="ignore",
        populate_by_name=True,
    )

    smtp_host: str
    smtp_port: int
    smtp_secure: bool = False
    smtp_user: str | None = None
    smtp_password: str | None = None
    smtp_from: str


def initialize_messaging(options: MessagingOptions) -> None:
    """Initialise the Hermes messaging layer with explicit configuration.

    This is the primary initialisation path when configuration is passed
    programmatically (e.g. from a config object loaded at application startup).

    Calling this function more than once has no effect — the first call wins
    (idempotent).

    Args:
        options: Messaging transport configuration.

    Example:
        >>> from hermes import initialize_messaging
        >>> from hermes.options import MessagingOptions
        >>> initialize_messaging(
        ...     MessagingOptions(
        ...         host="smtp.example.com",
        ...         port=587,
        ...         from_address="noreply@example.com",
        ...     )
        ... )
    """
    global _service
    if _service is not None:
        return
    _service = SmtpMessagingService(options)
    logger.debug("Hermes messaging layer initialised (explicit config).")


def initialize_messaging_from_env() -> None:
    """Initialise the Hermes messaging layer from environment variables.

    Reads the following variables (sourced from ``config.yaml`` via the
    service's startup config loader):

    +-----------------+----------+---------------------------------------------------+
    | Variable        | Required | Description                                       |
    +=================+==========+===================================================+
    | ``SMTP_HOST``   | ✓        | SMTP server hostname                              |
    +-----------------+----------+---------------------------------------------------+
    | ``SMTP_PORT``   | ✓        | SMTP server port                                  |
    +-----------------+----------+---------------------------------------------------+
    | ``SMTP_SECURE`` | –        | ``"true"`` to enable TLS (default: ``"false"``)   |
    +-----------------+----------+---------------------------------------------------+
    | ``SMTP_USER``   | –        | SMTP username; omit for unauthenticated relays    |
    +-----------------+----------+---------------------------------------------------+
    |``SMTP_PASSWORD``| –        | SMTP password; omit for unauthenticated relays    |
    +-----------------+----------+---------------------------------------------------+
    | ``SMTP_FROM``   | ✓        | Sender ``From`` address                           |
    +-----------------+----------+---------------------------------------------------+

    Calling this function more than once has no effect (idempotent).

    Raises:
        pydantic_core.ValidationError: If any required environment variable is
            missing or cannot be coerced to the expected type.

    Example:
        >>> from hermes import initialize_messaging_from_env
        >>> initialize_messaging_from_env()  # call once at application startup
    """
    global _service
    if _service is not None:
        return
    cfg = _SmtpEnvConfig()  # type: ignore[call-arg]
    _service = SmtpMessagingService(
        MessagingOptions(
            host=cfg.smtp_host,
            port=cfg.smtp_port,
            secure=cfg.smtp_secure,
            user=cfg.smtp_user,
            password=cfg.smtp_password,
            from_address=cfg.smtp_from,
        )
    )
    logger.debug("Hermes messaging layer initialised from environment.")


def get_messaging_service() -> IMessagingService:
    """Return the initialised messaging service singleton.

    Returns:
        The active :class:`~hermes.interfaces.messaging_service.IMessagingService`
        instance.

    Raises:
        RuntimeError: If neither :func:`initialize_messaging` nor
            :func:`initialize_messaging_from_env` has been called.

    Example:
        >>> from hermes import get_messaging_service
        >>> from hermes.interfaces import Message
        >>> messaging = get_messaging_service()
        >>> messaging.send(Message(to="ops@example.com", subject="Alert", body="Something failed."))
    """
    if _service is None:
        raise RuntimeError(
            "Hermes has not been initialised. "
            "Call initialize_messaging() or initialize_messaging_from_env() "
            "before using get_messaging_service()."
        )
    return _service


def _reset_messaging_for_testing() -> None:
    """Reset the internal messaging singleton.

    **For testing only.**  Call this in ``setup`` / ``teardown`` fixtures to
    ensure test isolation when testing code that calls
    :func:`initialize_messaging` or :func:`initialize_messaging_from_env`.
    """
    global _service
    _service = None

File: shared/Hermes/v1/hermes_ts/src/messaging.ts
import { z } from 'zod';
import type { IMessagingService } from './interfaces/messaging-service.js';
import { SmtpMessagingService } from './services/smtp-messaging-service.js';

/**
 * Configuration options for the Hermes messaging layer.
 *
 * Pass these options to {@link initializeMessaging} to configure the
 * messaging transport explicitly, or use {@link initializeMessagingFromEnv}
 * to load them automatically from environment variables.
 */
export interface MessagingOptions {
  /** SMTP server hostname (e.g. `'smtp.example.com'`). */
  host: string;

  /** SMTP server port (e.g. `587` for STARTTLS, `465` for SSL, `25` for plain). */
  port: number;

  /**
   * Whether to use TLS for the connection.
   * - `true` — wrap the connection in TLS from the start (port 465 / SMTPS).
   * - `false` — use plain or STARTTLS (ports 25 / 587).
   * Defaults to `false`.
   */
  secure?: boolean;

  /** SMTP authentication username. Omit for unauthenticated relays. */
  user?: string;

  /** SMTP authentication password. Omit for unauthenticated relays. */
  password?: string;

  /** The `From` address used for all outgoing messages (e.g. `'noreply@example.com'`). */
  from: string;
}

/** Zod schema for loading {@link MessagingOptions} from `process.env`. */
const messagingEnvSchema = z.object({
  SMTP_HOST: z.string().min(1, 'SMTP_HOST is required'),
  SMTP_PORT: z.coerce.number().int().min(1, 'SMTP_PORT must be a positive integer'),
  SMTP_SECURE: z
    .string()
    .optional()
    .transform((v) => v === 'true'),
  SMTP_USER: z.string().optional(),
  SMTP_PASSWORD: z.string().optional(),
  SMTP_FROM: z.string().min(1, 'SMTP_FROM is required'),
});

/** Singleton messaging service instance. */
let _service: IMessagingService | undefined;

/**
 * Initialises the Hermes messaging layer with explicit configuration.
 *
 * This is the primary initialisation path when configuration is passed
 * programmatically (e.g. from a config object loaded at application startup).
 *
 * Calling this function more than once has no effect — the first call wins
 * (idempotent).
 *
 * @param options - Messaging transport configuration.
 *
 * @example
 * ```typescript
 * import { initializeMessaging } from '@elastic-resume-base/hermes';
 *
 * initializeMessaging({
 *   host: process.env.SMTP_HOST ?? 'localhost',
 *   port: Number(process.env.SMTP_PORT) || 587,
 *   from: process.env.SMTP_FROM ?? 'noreply@example.com',
 * });
 * ```
 */
export function initializeMessaging(options: MessagingOptions): void {
  if (_service) return;
  _service = new SmtpMessagingService(options);
}

/**
 * Initialises the Hermes messaging layer from environment variables.
 *
 * Reads the following variables (all sourced from `config.yaml` via the
 * service's startup config loader):
 *
 * | Variable       | Required | Description                                      |
 * |----------------|----------|--------------------------------------------------|
 * | `SMTP_HOST`    | ✅       | SMTP server hostname                             |
 * | `SMTP_PORT`    | ✅       | SMTP server port                                 |
 * | `SMTP_SECURE`  | ❌       | `'true'` to enable TLS (default: `'false'`)      |
 * | `SMTP_USER`    | ❌       | SMTP username (omit for unauthenticated relays)  |
 * | `SMTP_PASSWORD`| ❌       | SMTP password (omit for unauthenticated relays)  |
 * | `SMTP_FROM`    | ✅       | Sender `From` address                            |
 *
 * Calling this function more than once has no effect (idempotent).
 *
 * @throws {ZodError} If any required environment variable is missing or invalid.
 *
 * @example
 * ```typescript
 * import { initializeMessagingFromEnv } from '@elastic-resume-base/hermes';
 *
 * // Call once at application startup — after config.yaml has been loaded.
 * initializeMessagingFromEnv();
 * ```
 */
export function initializeMessagingFromEnv(): void {
  if (_service) return;
  const env = messagingEnvSchema.parse(process.env);
  _service = new SmtpMessagingService({
    host: env.SMTP_HOST,
    port: env.SMTP_PORT,
    secure: env.SMTP_SECURE,
    user: env.SMTP_USER,
    password: env.SMTP_PASSWORD,
    from: env.SMTP_FROM,
  });
}

/**
 * Returns the initialised messaging service singleton.
 *
 * @returns The active {@link IMessagingService} instance.
 * @throws {Error} If neither {@link initializeMessaging} nor
 *   {@link initializeMessagingFromEnv} has been called.
 *
 * @example
 * ```typescript
 * import { getMessagingService } from '@elastic-resume-base/hermes';
 *
 * const messaging = getMessagingService();
 * await messaging.send({ to: 'ops@example.com', subject: 'Alert', body: 'Something failed.' });
 * ```
 */
export function getMessagingService(): IMessagingService {
  if (!_service) {
    throw new Error(
      'Hermes has not been initialised. ' +
        'Call initializeMessaging() or initializeMessagingFromEnv() before using getMessagingService().',
    );
  }
  return _service;
}

/**
 * Resets the internal messaging singleton.
 *
 * **For testing only.** Call this in `afterEach` / `beforeEach` blocks to
 * ensure test isolation when testing code that calls `initializeMessaging`.
 *
 * @internal
 */
export function _resetMessagingForTesting(): void {
  _service = undefined;
}
